File: book/lib/services/automl_service.dart
import 'api_client.dart';

/// Task types for ML
enum TaskType {
  classification,
  regression,
  clustering,
  dimensionalityReduction,
  anomalyDetection,
  timeSeries,
  nlp,
  computerVision,
  recommendation,
  reinforcementLearning,
}

/// Algorithm categories
enum AlgorithmCategory {
  linear,
  treeBased,
  ensemble,
  neuralNetwork,
  svm,
  neighbors,
  clustering,
  bayesian,
  dimensionality,
  boosting,
  deepLearning,
  transformer,
  other,
}

/// ML Algorithm
class Algorithm {
  final String id;
  final String name;
  final String category;
  final List<String> taskTypes;
  final String description;
  final List<Map<String, dynamic>> hyperparameters;
  final List<String> pros;
  final List<String> cons;
  final String complexity;
  final bool gpuAccelerated;
  final String library;

  Algorithm({
    required this.id,
    required this.name,
    required this.category,
    required this.taskTypes,
    required this.description,
    required this.hyperparameters,
    required this.pros,
    required this.cons,
    required this.complexity,
    required this.gpuAccelerated,
    required this.library,
  });

  factory Algorithm.fromJson(Map<String, dynamic> json) {
    return Algorithm(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      category: json['category'] ?? '',
      taskTypes: List<String>.from(json['task_types'] ?? []),
      description: json['description'] ?? '',
      hyperparameters: List<Map<String, dynamic>>.from(json['hyperparameters'] ?? []),
      pros: List<String>.from(json['pros'] ?? []),
      cons: List<String>.from(json['cons'] ?? []),
      complexity: json['complexity'] ?? '',
      gpuAccelerated: json['gpu_accelerated'] ?? false,
      library: json['library'] ?? '',
    );
  }
}

/// Model score
class ModelScore {
  final double? accuracy;
  final double? precision;
  final double? recall;
  final double? f1;
  final double? rocAuc;
  final double? mse;
  final double? rmse;
  final double? mae;
  final double? r2;
  final double? silhouette;
  final double? calinskiHarabasz;
  final double? daviesBouldin;

  ModelScore({
    this.accuracy,
    this.precision,
    this.recall,
    this.f1,
    this.rocAuc,
    this.mse,
    this.rmse,
    this.mae,
    this.r2,
    this.silhouette,
    this.calinskiHarabasz,
    this.daviesBouldin,
  });

  factory ModelScore.fromJson(Map<String, dynamic> json) {
    return ModelScore(
      accuracy: json['accuracy']?.toDouble(),
      precision: json['precision']?.toDouble(),
      recall: json['recall']?.toDouble(),
      f1: json['f1']?.toDouble(),
      rocAuc: json['roc_auc']?.toDouble(),
      mse: json['mse']?.toDouble(),
      rmse: json['rmse']?.toDouble(),
      mae: json['mae']?.toDouble(),
      r2: json['r2']?.toDouble(),
      silhouette: json['silhouette']?.toDouble(),
      calinskiHarabasz: json['calinski_harabasz']?.toDouble(),
      daviesBouldin: json['davies_bouldin']?.toDouble(),
    );
  }

  Map<String, double?> get classificationScores => {
    'Accuracy': accuracy,
    'Precision': precision,
    'Recall': recall,
    'F1': f1,
    'ROC-AUC': rocAuc,
  };

  Map<String, double?> get regressionScores => {
    'R²': r2,
    'MSE': mse,
    'RMSE': rmse,
    'MAE': mae,
  };

  Map<String, double?> get clusteringScores => {
    'Silhouette': silhouette,
    'Calinski-Harabasz': calinskiHarabasz,
    'Davies-Bouldin': daviesBouldin,
  };
}

/// Trained model
class TrainedModel {
  final String id;
  final String algorithmId;
  final String algorithmName;
  final Map<String, dynamic> hyperparameters;
  final ModelScore scores;
  final double trainingTime;
  final DateTime createdAt;
  final List<String> recommendations;

  TrainedModel({
    required this.id,
    required this.algorithmId,
    required this.algorithmName,
    required this.hyperparameters,
    required this.scores,
    required this.trainingTime,
    required this.createdAt,
    required this.recommendations,
  });

  factory TrainedModel.fromJson(Map<String, dynamic> json) {
    return TrainedModel(
      id: json['id'] ?? '',
      algorithmId: json['algorithm_id'] ?? '',
      algorithmName: json['algorithm_name'] ?? '',
      hyperparameters: Map<String, dynamic>.from(json['hyperparameters'] ?? {}),
      scores: ModelScore.fromJson(json['scores'] ?? {}),
      trainingTime: (json['training_time'] ?? 0).toDouble(),
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      recommendations: List<String>.from(json['recommendations'] ?? []),
    );
  }
}

/// Dataset info
class DatasetInfo {
  final String name;
  final int nSamples;
  final int nFeatures;
  final int? nClasses;
  final String targetColumn;
  final List<String> featureColumns;
  final String taskType;
  final bool hasMissing;
  final bool hasCategorical;

  DatasetInfo({
    required this.name,
    required this.nSamples,
    required this.nFeatures,
    this.nClasses,
    required this.targetColumn,
    required this.featureColumns,
    required this.taskType,
    required this.hasMissing,
    required this.hasCategorical,
  });

  factory DatasetInfo.fromJson(Map<String, dynamic> json) {
    return DatasetInfo(
      name: json['name'] ?? '',
      nSamples: json['n_samples'] ?? 0,
      nFeatures: json['n_features'] ?? 0,
      nClasses: json['n_classes'],
      targetColumn: json['target_column'] ?? '',
      featureColumns: List<String>.from(json['feature_columns'] ?? []),
      taskType: json['task_type'] ?? '',
      hasMissing: json['has_missing'] ?? false,
      hasCategorical: json['has_categorical'] ?? false,
    );
  }
}

/// AutoML Experiment
class AutoMLExperiment {
  final String id;
  final String name;
  final DatasetInfo datasetInfo;
  final String taskType;
  final List<String> algorithmsToTry;
  final String optimizationMetric;
  final int cvFolds;
  final int? maxTimeMinutes;
  final String status;
  final List<TrainedModel> models;
  final String? bestModelId;
  final DateTime createdAt;
  final DateTime? completedAt;

  AutoMLExperiment({
    required this.id,
    required this.name,
    required this.datasetInfo,
    required this.taskType,
    required this.algorithmsToTry,
    required this.optimizationMetric,
    required this.cvFolds,
    this.maxTimeMinutes,
    required this.status,
    required this.models,
    this.bestModelId,
    required this.createdAt,
    this.completedAt,
  });

  factory AutoMLExperiment.fromJson(Map<String, dynamic> json) {
    return AutoMLExperiment(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      datasetInfo: DatasetInfo.fromJson(json['dataset_info'] ?? {}),
      taskType: json['task_type'] ?? '',
      algorithmsToTry: List<String>.from(json['algorithms_to_try'] ?? []),
      optimizationMetric: json['optimization_metric'] ?? '',
      cvFolds: json['cv_folds'] ?? 5,
      maxTimeMinutes: json['max_time_minutes'],
      status: json['status'] ?? 'pending',
      models: (json['models'] as List?)?.map((m) => TrainedModel.fromJson(m)).toList() ?? [],
      bestModelId: json['best_model_id'],
      createdAt: DateTime.tryParse(json['created_at'] ?? '') ?? DateTime.now(),
      completedAt: json['completed_at'] != null ? DateTime.tryParse(json['completed_at']) : null,
    );
  }

  TrainedModel? get bestModel {
    if (bestModelId == null) return null;
    return models.firstWhere((m) => m.id == bestModelId, orElse: () => models.first);
  }
}

/// Algorithm recommendation
class AlgorithmRecommendation {
  final Algorithm algorithm;
  final int score;
  final List<String> reasons;

  AlgorithmRecommendation({
    required this.algorithm,
    required this.score,
    required this.reasons,
  });

  factory AlgorithmRecommendation.fromJson(Map<String, dynamic> json) {
    return AlgorithmRecommendation(
      algorithm: Algorithm.fromJson(json['algorithm'] ?? {}),
      score: json['score'] ?? 0,
      reasons: List<String>.from(json['reasons'] ?? []),
    );
  }
}

/// AutoML Service
class AutoMLService {
  final ApiClient _api;

  AutoMLService({ApiClient? api}) : _api = api ?? apiClient;

  // =========================================================================
  // ALGORITHMS
  // =========================================================================

  Future<List<Algorithm>> listAlgorithms({
    String? taskType,
    String? category,
    bool gpuOnly = false,
  }) async {
    var endpoint = '/api/automl/algorithms';
    final params = <String>[];

    if (taskType != null) params.add('task_type=$taskType');
    if (category != null) params.add('category=$category');
    if (gpuOnly) params.add('gpu_only=true');

    if (params.isNotEmpty) {
      endpoint += '?${params.join('&')}';
    }

    final data = await _api.getList(endpoint);
    return data.map((json) => Algorithm.fromJson(json)).toList();
  }

  Future<Algorithm> getAlgorithm(String algorithmId) async {
    final data = await _api.get('/api/automl/algorithms/$algorithmId');
    return Algorithm.fromJson(data);
  }

  Future<List<Algorithm>> getAlgorithmsForTask(String taskType) async {
    final data = await _api.getList('/api/automl/algorithms/task/$taskType');
    return data.map((json) => Algorithm.fromJson(json)).toList();
  }

  Future<List<Map<String, dynamic>>> listCategories() async {
    final data = await _api.getList('/api/automl/categories');
    return data.cast<Map<String, dynamic>>();
  }

  Future<Map<String, dynamic>> listTaskTypes() async {
    return await _api.get('/api/automl/task-types');
  }

  // =========================================================================
  // EXPERIMENTS
  // =========================================================================

  Future<AutoMLExperiment> createExperiment({
    required String name,
    required String datasetPath,
    required String targetColumn,
    required String taskType,
    List<String>? algorithms,
    String? optimizationMetric,
    int cvFolds = 5,
    int? maxTimeMinutes = 60,
    double testSize = 0.2,
  }) async {
    final data = await _api.post('/api/automl/experiments', {
      'name': name,
      'dataset_path': datasetPath,
      'target_column': targetColumn,
      'task_type': taskType,
      'algorithms': algorithms,
      'optimization_metric': optimizationMetric,
      'cv_folds': cvFolds,
      'max_time_minutes': maxTimeMinutes,
      'test_size': testSize,
    });
    return AutoMLExperiment.fromJson(data);
  }

  Future<List<AutoMLExperiment>> listExperiments() async {
    final data = await _api.getList('/api/automl/experiments');
    return data.map((json) => AutoMLExperiment.fromJson(json)).toList();
  }

  Future<AutoMLExperiment> getExperiment(String experimentId) async {
    final data = await _api.get('/api/automl/experiments/$experimentId');
    return AutoMLExperiment.fromJson(data);
  }

  Future<void> deleteExperiment(String experimentId) async {
    await _api.delete('/api/automl/experiments/$experimentId');
  }

  Future<void> stopExperiment(String experimentId) async {
    await _api.post('/api/automl/experiments/$experimentId/stop', {});
  }

  Future<List<TrainedModel>> getExperimentModels(String experimentId) async {
    final data = await _api.getList('/api/automl/experiments/$experimentId/models');
    return data.map((json) => TrainedModel.fromJson(json)).toList();
  }

  Future<TrainedModel> getBestModel(String experimentId) async {
    final data = await _api.get('/api/automl/experiments/$experimentId/best-model');
    return TrainedModel.fromJson(data);
  }

  // =========================================================================
  // RECOMMENDATIONS
  // =========================================================================

  Future<List<AlgorithmRecommendation>> getRecommendations({
    required String taskType,
    required int nSamples,
    required int nFeatures,
    bool hasCategorical = false,
    bool hasMissing = false,
    bool needInterpretability = false,
    bool needSpeed = false,
    bool hasGpu = false,
  }) async {
    final data = await _api.post('/api/automl/recommend', {
      'task_type': taskType,
      'n_samples': nSamples,
      'n_features': nFeatures,
      'has_categorical': hasCategorical,
      'has_missing': hasMissing,
      'need_interpretability': needInterpretability,
      'need_speed': needSpeed,
      'has_gpu': hasGpu,
    });

    final recommendations = data['recommendations'] as List? ?? [];
    return recommendations.map((r) => AlgorithmRecommendation.fromJson(r)).toList();
  }

  // =========================================================================
  // QUICK TRAIN
  // =========================================================================

  Future<Map<String, dynamic>> quickTrain({
    required String algorithmId,
    required String datasetPath,
    required String targetColumn,
    Map<String, dynamic>? hyperparameters,
    double testSize = 0.2,
  }) async {
    return await _api.post('/api/automl/quick-train', {
      'algorithm_id': algorithmId,
      'dataset_path': datasetPath,
      'target_column': targetColumn,
      'hyperparameters': hyperparameters,
      'test_size': testSize,
    });
  }
}

final automlService = AutoMLService();

File: notebook/src/services/automlService.ts
// AutoML Service - Automated Machine Learning experiments

import apiClient from "./apiClient";

export type TaskType =
  | "classification"
  | "regression"
  | "clustering"
  | "dimensionality_reduction"
  | "anomaly_detection"
  | "time_series";

export type AlgorithmCategory =
  | "linear"
  | "tree_based"
  | "ensemble"
  | "neural_network"
  | "svm"
  | "neighbors"
  | "clustering"
  | "bayesian"
  | "boosting";

export interface Algorithm {
  id: string;
  name: string;
  category: AlgorithmCategory;
  taskTypes: TaskType[];
  description: string;
  hyperparameters: Array<{
    name: string;
    type: string;
    default?: unknown;
    min?: number;
    max?: number;
    options?: string[];
  }>;
  pros: string[];
  cons: string[];
  complexity: string;
  gpuAccelerated: boolean;
  library: string;
}

export interface ModelScore {
  accuracy?: number;
  precision?: number;
  recall?: number;
  f1?: number;
  rocAuc?: number;
  mse?: number;
  rmse?: number;
  mae?: number;
  r2?: number;
  silhouette?: number;
}

export interface TrainedModel {
  id: string;
  algorithmId: string;
  algorithmName: string;
  hyperparameters: Record<string, unknown>;
  scores: ModelScore;
  trainingTime: number;
  createdAt: Date;
  recommendations: string[];
}

export interface AutoMLExperiment {
  id: string;
  name: string;
  datasetPath: string;
  targetColumn: string;
  taskType: TaskType;
  algorithmsToTry: string[];
  optimizationMetric: string;
  cvFolds: number;
  maxTimeMinutes?: number;
  testSize: number;
  status: "pending" | "running" | "completed" | "failed" | "stopped";
  progress: number;
  models: TrainedModel[];
  bestModelId?: string;
  createdAt: Date;
  completedAt?: Date;
  error?: string;
}

export interface AlgorithmRecommendation {
  algorithm: Algorithm;
  score: number;
  reasons: string[];
}

export interface CreateExperimentParams {
  name: string;
  datasetPath: string;
  targetColumn: string;
  taskType: TaskType;
  algorithms?: string[];
  optimizationMetric?: string;
  cvFolds?: number;
  maxTimeMinutes?: number;
  testSize?: number;
}

interface AlgorithmResponse {
  id: string;
  name: string;
  category: string;
  task_types: string[];
  description: string;
  hyperparameters: Array<{
    name: string;
    type: string;
    default?: unknown;
    min?: number;
    max?: number;
    options?: string[];
  }>;
  pros: string[];
  cons: string[];
  complexity: string;
  gpu_accelerated: boolean;
  library: string;
}

interface ExperimentResponse {
  id: string;
  name: string;
  dataset_path: string;
  target_column: string;
  task_type: string;
  algorithms_to_try: string[];
  optimization_metric: string;
  cv_folds: number;
  max_time_minutes?: number;
  test_size: number;
  status: string;
  progress: number;
  models: Array<{
    id: string;
    algorithm_id: string;
    algorithm_name: string;
    hyperparameters: Record<string, unknown>;
    scores: Record<string, number>;
    training_time: number;
    created_at: string;
    recommendations: string[];
  }>;
  best_model_id?: string;
  created_at: string;
  completed_at?: string;
  error?: string;
}

class AutoMLService {
  private parseAlgorithm(data: AlgorithmResponse): Algorithm {
    return {
      id: data.id,
      name: data.name,
      category: data.category as AlgorithmCategory,
      taskTypes: data.task_types as TaskType[],
      description: data.description,
      hyperparameters: data.hyperparameters,
      pros: data.pros,
      cons: data.cons,
      complexity: data.complexity,
      gpuAccelerated: data.gpu_accelerated,
      library: data.library,
    };
  }

  private parseExperiment(data: ExperimentResponse): AutoMLExperiment {
    return {
      id: data.id,
      name: data.name,
      datasetPath: data.dataset_path,
      targetColumn: data.target_column,
      taskType: data.task_type as TaskType,
      algorithmsToTry: data.algorithms_to_try,
      optimizationMetric: data.optimization_metric,
      cvFolds: data.cv_folds,
      maxTimeMinutes: data.max_time_minutes,
      testSize: data.test_size,
      status: data.status as AutoMLExperiment["status"],
      progress: data.progress,
      models: data.models.map((m) => ({
        id: m.id,
        algorithmId: m.algorithm_id,
        algorithmName: m.algorithm_name,
        hyperparameters: m.hyperparameters,
        scores: m.scores as ModelScore,
        trainingTime: m.training_time,
        createdAt: new Date(m.created_at),
        recommendations: m.recommendations,
      })),
      bestModelId: data.best_model_id,
      createdAt: new Date(data.created_at),
      completedAt: data.completed_at ? new Date(data.completed_at) : undefined,
      error: data.error,
    };
  }

  // Algorithms
  async getAlgorithms(): Promise<Algorithm[]> {
    const response = await apiClient.get<AlgorithmResponse[]>("/api/automl/algorithms");
    return response.map((a) => this.parseAlgorithm(a));
  }

  async getAlgorithm(id: string): Promise<Algorithm> {
    const response = await apiClient.get<AlgorithmResponse>(`/api/automl/algorithms/${id}`);
    return this.parseAlgorithm(response);
  }

  async getAlgorithmsByTask(taskType: TaskType): Promise<Algorithm[]> {
    const response = await apiClient.get<AlgorithmResponse[]>(
      `/api/automl/algorithms?task_type=${taskType}`
    );
    return response.map((a) => this.parseAlgorithm(a));
  }

  // Experiments
  async listExperiments(): Promise<AutoMLExperiment[]> {
    const response = await apiClient.get<ExperimentResponse[]>("/api/automl/experiments");
    return response.map((e) => this.parseExperiment(e));
  }

  async getExperiment(id: string): Promise<AutoMLExperiment> {
    const response = await apiClient.get<ExperimentResponse>(`/api/automl/experiments/${id}`);
    return this.parseExperiment(response);
  }

  async createExperiment(params: CreateExperimentParams): Promise<AutoMLExperiment> {
    const response = await apiClient.post<ExperimentResponse>("/api/automl/experiments", {
      name: params.name,
      dataset_path: params.datasetPath,
      target_column: params.targetColumn,
      task_type: params.taskType,
      algorithms: params.algorithms,
      optimization_metric: params.optimizationMetric,
      cv_folds: params.cvFolds,
      max_time_minutes: params.maxTimeMinutes,
      test_size: params.testSize,
    });
    return this.parseExperiment(response);
  }

  async stopExperiment(id: string): Promise<void> {
    await apiClient.post(`/api/automl/experiments/${id}/stop`, {});
  }

  async deleteExperiment(id: string): Promise<void> {
    await apiClient.delete(`/api/automl/experiments/${id}`);
  }

  async rerunExperiment(id: string): Promise<AutoMLExperiment> {
    const response = await apiClient.post<ExperimentResponse>(
      `/api/automl/experiments/${id}/rerun`,
      {}
    );
    return this.parseExperiment(response);
  }

  // Recommendations
  async getRecommendations(params: {
    taskType: TaskType;
    nSamples: number;
    nFeatures: number;
    hasCategorical?: boolean;
    hasMissing?: boolean;
    needInterpretability?: boolean;
    needSpeed?: boolean;
    hasGpu?: boolean;
  }): Promise<AlgorithmRecommendation[]> {
    const response = await apiClient.post<
      Array<{
        algorithm: AlgorithmResponse;
        score: number;
        reasons: string[];
      }>
    >("/api/automl/recommendations", {
      task_type: params.taskType,
      n_samples: params.nSamples,
      n_features: params.nFeatures,
      has_categorical: params.hasCategorical,
      has_missing: params.hasMissing,
      need_interpretability: params.needInterpretability,
      need_speed: params.needSpeed,
      has_gpu: params.hasGpu,
    });

    return response.map((r) => ({
      algorithm: this.parseAlgorithm(r.algorithm),
      score: r.score,
      reasons: r.reasons,
    }));
  }

  // Model operations
  async exportModel(experimentId: string, modelId: string, format: "pickle" | "onnx" | "joblib"): Promise<Blob> {
    const response = await fetch(
      `${apiClient.getBaseUrl()}/api/automl/experiments/${experimentId}/models/${modelId}/export?format=${format}`
    );
    return response.blob();
  }

  async predictWithModel(
    experimentId: string,
    modelId: string,
    data: Record<string, unknown>[]
  ): Promise<unknown[]> {
    const response = await apiClient.post<{ predictions: unknown[] }>(
      `/api/automl/experiments/${experimentId}/models/${modelId}/predict`,
      { data }
    );
    return response.predictions;
  }

  // Dataset analysis
  async analyzeDataset(path: string): Promise<{
    nSamples: number;
    nFeatures: number;
    columns: Array<{
      name: string;
      dtype: string;
      unique: number;
      missing: number;
      isCategorical: boolean;
    }>;
    suggestedTaskType: TaskType;
    suggestedTarget: string;
  }> {
    const response = await apiClient.post<{
      n_samples: number;
      n_features: number;
      columns: Array<{
        name: string;
        dtype: string;
        unique: number;
        missing: number;
        is_categorical: boolean;
      }>;
      suggested_task_type: string;
      suggested_target: string;
    }>("/api/automl/analyze-dataset", { path });

    return {
      nSamples: response.n_samples,
      nFeatures: response.n_features,
      columns: response.columns.map((c) => ({
        name: c.name,
        dtype: c.dtype,
        unique: c.unique,
        missing: c.missing,
        isCategorical: c.is_categorical,
      })),
      suggestedTaskType: response.suggested_task_type as TaskType,
      suggestedTarget: response.suggested_target,
    };
  }
}

export const automlService = new AutoMLService();
export default automlService;
